lib/ace/ext/papeeria/katex-previewer.coffee

```diff
@@ -50,43 +55,129 @@ define((require, exports, module) ->
       )
       return
 
-    callbackHidePopover = ->
-      popoverHandler.hide($("#formula"))
-      editor.off("changeSelection", callbackHidePopover)
-      editor.session.off("changeScrollTop", callbackHidePopover)
-      editor.session.off("changeScrollLeft", callbackHidePopover)
-      return
+    jqEditorContainer = $(editor.container)
+    jqFormula = -> $("#formula")
+
+    # ch stands for Context Handler
+    contextHandler = new class
+      @removeRegex: /\\end\{equation\}|\\begin\{equation\}|\\label\{[^\}]*\}/g
+
+      @getEquationRange: (cursorRow) ->
+        i = cursorRow
+        while getContext(editor.session, i - 1) == "equation"
+          i -= 1
+        start = i
+        while getContext(editor.session, i + 1) == "equation"
+          i += 1
+        end = i
+        return [start, end]
+
+      @getWholeEquation: (start, end) ->
+        editor.session.getLines(start, end).join(" ").replace(@removeRegex, "")
+
+      @getPopoverPosition: (row) -> {
+          top: "#{editor.renderer.textToScreenCoordinates(row + 2, 1).pageY}px"
+          left: "#{jqEditorContainer.position().left}px"
+        }
+
+      constructor: ->
+        [@curStart, @curEnd] = [null, null]
```

Comment:
Well, they are also used in `delayedUpdatePopover` + `updatePopover` + `getCurrentFormula`, and `updatePosition` + `getPopoverPosition`. And the problem is, both of these are used as callback function [here](https://github.com/bardsoftware/ace/pull/72/files#diff-30dc2f6524f3c4b7209018a9fdcc526bR129). If any arguments had to be passed to `delayedUpdateFormula` or `updatePosition`, that would really complicate, how we pass them as callbacks, because we have to pass them to `editor.off` [here](https://github.com/bardsoftware/ace/pull/72/files#diff-30dc2f6524f3c4b7209018a9fdcc526bR133).